:heavy_plus_sign: Add new chapter DevOps (#267)
Il-Libro-Open-Source/book: docs/it/cultura/devops.md

---
layout: default
title: DevOps in pillole
parent: Cultura
nav_order: 1
---

<!-- prettier-ignore-start -->
# DevOps in pillole
{: .no_toc }

- TOC
{:toc}

<!-- prettier-ignore-end -->

## DevOps: Una rivoluzione nella cultura del software

> DevOps rappresenta molto più di una semplice metodologia tecnica: è un vero e proprio cambiamento culturale che coinvolge l’intera organizzazione, ridefinendo il modo in cui sviluppo (_development_) e operazioni (_operations_) collaborano per realizzare e distribuire software di qualità, in modo rapido ed efficiente. In un mercato sempre più competitivo e dinamico, aziende come Google, Amazon, Facebook e Netflix hanno abbracciato DevOps per garantire innovazione continua e risposte tempestive alle esigenze degli utenti.

Reviewer: Necessario riferimento, perché è un'affermazione piuttosto generica

Reviewer 2: Condivido il pensiero di @serenasensini

## I principi fondamentali di DevOps

Alla base di DevOps ci sono alcuni principi chiave che guidano i team verso prestazioni elevate:

> - **Collaborazione continua:** DevOps elimina i silos tradizionali tra sviluppo e operations, promuovendo un ambiente in cui la comunicazione è aperta e i team lavorano insieme durante tutto il ciclo di vita dell’applicazione.

Reviewer: AI-gen?

Reviewer 2: Sarebbe l'ideale definire "silos" ed "operations" per i neofiti

- **Automazione:** L’automazione dei processi, dalla scrittura del codice al testing, fino al rilascio e al monitoraggio, riduce la complessità, minimizza gli errori umani e accelera i tempi di consegna.
- **Integrazione e consegna continue (CI/CD):** Questi processi consentono di integrare regolarmente le modifiche al codice e di distribuirle rapidamente in ambienti di produzione, garantendo feedback immediati e una maggiore affidabilità del software.
- **Infrastruttura come codice (IaC):** Gestire l’infrastruttura tramite codice permette di replicare ambienti in modo coerente e di automatizzare il provisioning delle risorse.
- **Feedback rapido e miglioramento continuo:** Il monitoraggio costante e i cicli di feedback brevi permettono ai team di individuare e risolvere rapidamente i problemi, adattando il prodotto alle reali esigenze degli utenti.

### Cultura della Collaborazione e della Condivisione

**Il cuore del DevOps è la cultura**. Nessuno strumento può compensare la mancanza di comunicazione tra team.

**Best Practice:**

> - _Team cross-funzionali_: sviluppatori, sistemisti, QA e sicurezza devono collaborare fin dalle fasi iniziali del progetto.

Reviewer: ```suggestion
- _Team cross-funzionali_: chi sviluppa, chi si occupa di sysadmin o di QA e sicurezza deve collaborare fin dalle fasi iniziali del progetto.
```

- _Riunioni quotidiane brevi (stand-up)_: migliorano la visibilità e la sincronizzazione del lavoro.
- _Blameless post-mortem_: analizzare gli incidenti senza cercare colpevoli, ma per imparare e migliorare.

### Integrazione e Consegna Continua (CI/CD)

Automatizzare build, test e deploy riduce gli errori e accelera le iterazioni.

**Best Practice:**

- _Pipeline CI/CD automatizzate_: con strumenti come GitHub Actions, GitLab CI, Jenkins, CircleCI.

> - _Test automatici e di regressione_: unit, integration, end-to-end.

Reviewer: Sarebbe utile citare il capitolo dedicato al testing, visto che qui non si è approfondito molto: https://github.com/Il-Libro-Open-Source/book/blob/main/docs/it/sviluppo/testing.md

Reviewer 2: Condivido @serenasensini

- _Feature toggle_: per rilasciare codice in produzione in modo sicuro senza esporre subito tutte le funzionalità.

### Infrastructure as Code (IaC)

L’infrastruttura è parte integrante del software e deve essere trattata come codice.

**Best Practice:

> **- Uso di strumenti come Terraform, Ansible, Pulumi o AWS CloudFormation.

Reviewer: ```suggestion
**
```

> 

Reviewer: ```suggestion
**- Uso di strumenti come Terraform, Ansible, Pulumi o AWS CloudFormation.
```

> - Versionamento dell’infrastruttura in git, con revisioni e code review.

Reviewer: ```suggestion
- Versionamento dell’infrastruttura tramite git, con revisioni e code review.
```

> - Ambienti replicabili: staging identico alla produzione per test più affidabili.

Reviewer: Affermazione da motivare. Staging identico alla produzione? Delle differenze sostanziali ci sono... Pre-prod e staging sono lo stesso ambiente, questo sì, ma una differenza è che l'ambiente di sviluppo viene aggiornato frequentemente con funzionalità che raggiungeranno la fase di produzione al termine di una release di produzione. Questo permette di testarle e di familiarizzarsi con esse prima della release di produzione. Se sono perfettamente identici, è perché normalmente siamo in una fase di pre-rilascio, o prima del go-live

### Monitoraggio e logging continui

> DevOps senza visibilità è cieco. Il monitoraggio è essenziale per garantire affidabilità e performance.

Reviewer: ```suggestion
**DevOps senza visibilità è cieco**. Il monitoraggio è essenziale per garantire affidabilità e performance.
```
Rivedrei questa frase

**Best Practice:**

> - Monitoraggio proattivo: con alert su metriche critiche (CPU, RAM, errori, latenza).

Reviewer: Generico, argomentare meglio

- Log centralizzati e strutturati: tramite ELK stack, Grafana Loki, o servizi cloud (CloudWatch, Azure Monitor).

> - SLO/SLA e SLI: definizione chiara degli obiettivi di affidabilità e indicatori di performance.

Reviewer: Qui avrei dettagliato meglio il concetto di SLO/SLA, perché così sono fuori contesto

### Sicurezza DevSecOps

Integrare la sicurezza fin dall’inizio del ciclo di sviluppo, non alla fine.

> **Best Practice:**

Reviewer: AI-gen?

- Scan del codice statico (SAST) e delle dipendenze (SCA) in CI.

> - Gestione segreti sicura (Vault, AWS Secrets Manager, GitHub Actions Secrets).

Reviewer: ```suggestion
- Gestione segreti sicura (Vault, AWS Secrets Manager, GitHub Actions Secrets).
```
Dettagliare meglio cosa sono i segreti. Non diamo per scontato che chi legge sappia a che servono e cosa sono, né che conosca quegli strumenti

- Controlli di accesso basati sui ruoli (RBAC) e principio del minimo privilegio.

### Automazione intelligente

Automatizzare tutto ciò che è ripetibile, ma con criterio.

> **Best Practice:**

Reviewer: AI-gen?

- Automazione del provisioning, delle configurazioni, dei test e dei rilasci.
- ChatOps: integrazione dei comandi DevOps in strumenti come Slack o MS Teams.
- Documentazione automatica e aggiornamento continuo delle pipeline.

### Feedback rapido e iterazioni brevi

Il ciclo di feedback deve essere continuo e veloce, per favorire l'apprendimento e il miglioramento costante.

**Best Practice:**

- Release frequenti e a basso rischio.

> - Canary releases e blue/green deployment per evitare downtime.

Reviewer: Sarebbe utile spiegare il termine di deployment canary ancor prima di release

- Raccolta attiva dei feedback dagli utenti e dai sistemi.

## La cultura DevOps

DevOps non si limita all’adozione di nuovi strumenti o processi, ma richiede una trasformazione culturale profonda. I valori fondanti sono trasparenza, responsabilità condivisa e fiducia reciproca. Ogni membro del team è coinvolto sia nello sviluppo che nella gestione operativa, assumendosi la responsabilità dell’intero ciclo di vita del prodotto, dall’ideazione al rilascio e oltre.

> "DevOps è un cambiamento culturale in cui i team adottano una cultura di progettazione del software, un flusso di lavoro e un set di strumenti che elevano i requisiti operativi allo stesso livello di importanza rivestito dall'architettura, dalla progettazione e dallo sviluppo."

## Vantaggi di DevOps

Adottare DevOps porta benefici tangibili su più livelli:

- **Tecnici:** Rilasci più frequenti e affidabili, riduzione della complessità, risoluzione rapida dei problemi.
- **Culturali:** Team più produttivi, motivati e orientati al risultato, maggiore soddisfazione dei clienti.

> - **Aziendali:** Accelerazione del time-to-market, ambienti più stabili, migliore allineamento tra obiettivi di business e sviluppo tecnologico.

Reviewer: ```suggestion
- **Aziendali:** Accelerazione del time-to-market, ambienti più stabili, migliore allineamento tra obiettivi di business e sviluppo tecnologico.
```

Più stabili in che modo?

## DevOps e la sicurezza come si sposano?

Nel contesto DevOps, la sicurezza non può più essere un’attività isolata a fine ciclo.
Nasce così **DevSecOps**, un'estensione naturale del DevOps che ha l’obiettivo di integrare la sicurezza fin dall’inizio del processo di sviluppo, rendendola parte integrante della cultura e delle pipeline.

### DevSecOps in pratica

La sicurezza viene automatizzata insieme ai test e al deployment.
Gli sviluppatori hanno visibilità sui problemi di sicurezza durante il coding, grazie a tool di analisi statica (SAST) o di gestione delle dipendenze vulnerabili (SCA).
I team condividono la responsabilità della compliance, della gestione dei segreti e delle configurazioni sicure.
Le pipeline CI/CD includono controlli di sicurezza automatici, come scan delle immagini Docker o verifica delle policy infrastrutturali.

> DevSecOps punta a spostare la sicurezza “a sinistra” nel ciclo di sviluppo, garantendo che ogni cambiamento – dal codice all’infrastruttura – sia sicuro per impostazione predefinita, senza rallentare il rilascio del software.

Reviewer: ```suggestion
DevSecOps punta a spostare la sicurezza “nella fase di sinistra”, ossia nel ciclo di sviluppo, garantendo che ogni cambiamento – dal codice all’infrastruttura – sia sicuro per impostazione predefinita, senza rallentare il rilascio del software.
```

## Fonti

- https://www.atlassian.com/it/devops/what-is-devops
- https://github.com/andredesousa/devops-best-practices
- https://www.atlassian.com/it/devops/what-is-devops/benefits-of-devops
- https://www.apogeonline.com/libri/devops-fabio-mora/

> - https://about.gitlab.com/it-it/topics/devops//

Reviewer: ```suggestion
- https://about.gitlab.com/it-it/topics/devops/
```

---
layout: default
title: DevOps in pillole
parent: Cultura
nav_order: 1
---

<!-- prettier-ignore-start -->
# DevOps in pillole
{: .no_toc }

- TOC
{:toc}

<!-- prettier-ignore-end -->

# DevOps: Sicurezza e Automazione per Team Agili

Prima dell'avvento delle pratiche DevOps, lo sviluppo software seguiva un modello rigidamente compartimentato che in molte organizzazioni persiste ancora oggi.

> Gli sviluppatori scrivevano codice per settimane o mesi, poi lo passavano al team di QA, che a sua volta, dopo ulteriori settimane di test, lo passava al team operations per il deployment. Ogni step era un collo di bottiglia: configurazioni manuali su server, deployment notturni del venerdì sera con l'intero team in conference call, e l'inevitabile gioco dello "scaricabarile" quando qualcosa andava storto in produzione ("funzionava nel mio ambiente 😅").

Reviewer: ```suggestion
Il team di sviluppo scriveva codice per settimane o mesi, poi lo passava al team di QA, che a sua volta, dopo ulteriori settimane di test, lo passava al team operations per il deployment. Ogni step era un collo di bottiglia: configurazioni manuali su server, deployment notturni del venerdì sera con l'intero team in conference call, e l'inevitabile gioco dello "scaricabarile" quando qualcosa andava storto in produzione ("funzionava nel mio ambiente 😅").
```

- **Developer Experience (DevEx)** come pratica trasversale: toolchain coerente, ambienti locali e preview veloci, automazione dell’“inner loop”, riduzione del carico cognitivo e piattaforme self‑service.

## La Sicurezza: DevSecOps senza overhead

DevSecOps integra le pratiche di sicurezza in tutti i processi DevOps, generando automaticamente artifact di sicurezza e compliance durante l'intero processo, inclusi sviluppo, build, packaging, distribuzione e deployment.

### Principi Fondamentali del DevSecOps

**Shift Left Security**: La sicurezza non è più una fase finale, ma inizia dal primo commit. Questo significa:

- Analisi di vulnerabilità nel codice durante la scrittura
- Scansione delle dipendenze in tempo reale
- Policy as Code (regole come codice) per definire regole di sicurezza

**Security by Design**: Ogni decisione architetturale considera l'impatto sulla sicurezza. Per team piccoli, questo si traduce in:

- Template di progetto con security baseline
- Checklist di sicurezza integrate nel workflow
- Automazione di controlli di compliance

## L'Esperienza: DevEx come moltiplicatore

> La Developer Experience (DevEx) rappresenta l'insieme delle percezioni, sentimenti e risposte che gli sviluppatori hanno mentre interagiscono con strumenti, processi, pratiche e culture nel loro ambiente di lavoro. Non è semplicemente una questione di tool migliori, ma un approccio sistemico che considera ogni punto di frizione nel percorso dello sviluppatore - dal momento in cui apre l'IDE fino al deploy in produzione.

Reviewer: ```suggestion
La Developer Experience (DevEx) rappresenta l'insieme delle percezioni, sentimenti e risposte che i/le developer hanno mentre interagiscono con strumenti, processi, pratiche e culture nel loro ambiente di lavoro. Non è semplicemente una questione di tool migliori, ma un approccio sistemico che considera ogni punto di frizione nel percorso di chi sviluppa - dal momento in cui apre l'IDE fino al deploy in produzione.
```

C'è chi, comunque, obietta sull'unire il contesto DevOps con la DevEx, e forse vale la pena fae una riflessione: https://www.youtube.com/watch?v=1g3NL6pOGCg&t=43

### Componenti Chiave della DevEx

La DevEx si fonda su tre dimensioni interconnesse che determinano la produttività e la soddisfazione degli sviluppatori:

**Flow State e Focus Time**: La capacità di entrare e mantenere uno stato di concentrazione profonda è fondamentale per il lavoro creativo dello sviluppo software. Questo richiede:

- Minimizzazione delle interruzioni e context switching
- Tool che rispondono istantaneamente senza breaking del flusso di pensiero
- Processi asincroni che non bloccano il lavoro in attesa di approvazioni

**Autonomia e Self-Service**: Gli sviluppatori devono poter operare indipendentemente senza dipendenze esterne eccessive:

- Accesso diretto agli ambienti di sviluppo e test
- Capacità di provisioning risorse senza ticket
- Documentazione e tool discoverable che non richiedono training estensivo

**Feedback Loop Quality**: La qualità e velocità del feedback determina quanto rapidamente uno sviluppatore può iterare e migliorare:

- Build e test che completano in minuti, non ore
- Error messages chiari e actionable che guidano verso la soluzione
- Metriche di performance accessibili che mostrano l'impatto del proprio lavoro

> Questi principi si traducono in pratiche concrete: ambienti di sviluppo standardizzati ma personalizzabili, pipeline CI/CD che "funzionano e basta", documentazione che vive accanto al codice, e tool che amplificano le capacità piuttosto che aggiungere complessità. Per team piccoli, una buona DevEx non è un lusso ma una necessità strategica - è ciò che permette a poche persone di competere con organizzazioni molto più grandi, eliminando friction e moltiplicando l'efficacia di ogni singolo sviluppatore.

Reviewer: ```suggestion
Questi principi si traducono in pratiche concrete: ambienti di sviluppo standardizzati ma personalizzabili, pipeline CI/CD che "funzionano e basta", documentazione che vive accanto al codice, e tool che amplificano le capacità piuttosto che aggiungere complessità. Per team piccoli, una buona DevEx non è un lusso ma una necessità strategica - è ciò che permette a poche persone di competere con organizzazioni molto più grandi, eliminando friction e moltiplicando l'efficacia di ogni singolo membro del team di sviluppo.
```

## I Pilastri: CALMS framework applicato

Il framework CALMS(Culture Automation Lean Measurement Sharing), coniato da Jez Humble co-autore di "The DevOps Handbook", valuta la capacità di un'azienda di adottare processi DevOps e rappresenta un modo per misurare il successo durante una trasformazione DevOps. L'acronimo sta per Culture, Automation, Lean, Measurement e Sharing.

### Applicazione Pratica del CALMS per Team Piccoli

**Culture**: In team piccoli, la cultura si forma rapidamente ma è fragile:

- Responsabilità condivisa del codice (no "not my code" mentality)
- Fail fast, learn fast: gli errori sono opportunità di apprendimento
- Transparency: tutti sanno su cosa stanno lavorando gli altri

**Automation**: L'automazione è il moltiplicatore di forza per team piccoli:

- Ogni processo ripetuto più di 3 volte deve essere automatizzato
- Pipeline che gestiscono tutto: da commit a produzione
- Self-service per operazioni comuni

**Lean**: Eliminare tutto ciò che non aggiunge valore:

- Processi semplici e lineari

> - Riduzione dei handoff tra persone

Reviewer: Aggiungerei una frase per spiegare il concetto di handoff, per chi non è interno al tema

    style A fill:#ffcccc
    style L fill:#ccffcc
    style K fill:#ff9999
    style U fill:#99ff99
    style F fill:#ff9999
    style V fill:#ffcc99
```

Ogni commit e deploy viene automaticamente validato contro questi criteri.
Se non rispetta gli standard, il processo si ferma - nessun intervento umano necessario.
Questo è Security as Code: policy di sicurezza che si auto-applicano attraverso automazione, non attraverso documenti o checklist manuali.

### Fase 2: Security Integrata e continua

Durante lo sviluppo, la sicurezza è continua e automatica:

```mermaid
sequenceDiagram
    participant Dev as Developer
    participant IDE as IDE/Editor
    participant VCS as Version Control
    participant CI as CI Pipeline

    Dev->>IDE: Scrive codice
    IDE->>Dev: Security linting real-time
    Dev->>VCS: Push commit
    VCS->>CI: Trigger pipeline
    CI->>CI: Security scan + tests
    CI->>Dev: Feedback immediato
```

**Strumenti Integrati nello Sviluppo**:

- Security linter nell'IDE che evidenzia problemi prima del commit
- Pre-commit hook che bloccano commit non sicuri
- Dependency vulnerability warnings in tempo reale

### Fase 3: Pipeline di Rilascio Automatizzate

> La pipeline trasforma il codice in valore business in modo sicuro e ripetibile:

Reviewer: Aggiungere dettaglio sul modo in cui l'automazione supporta questo flusso

5. **Deploy**: Deployment automatico con verification
6. **Monitor**: Health check e security monitoring

## La Misura: Come sapere se funziona

Post-deployment, il sistema fornisce feedback continuo attraverso metriche oggettive che indicano la salute del processo DevOps. Misurare non significa controllare, ma capire dove intervenire per migliorare.

### DORA Metrics: Il Gold Standard della Performance

Le metriche DORA (DevOps Research and Assessment) sono quattro indicatori chiave che predicono le performance di un'organizzazione software:

- **Deployment Frequency**: Quanto spesso rilasciamo in produzione
- **Lead Time for Changes**: Tempo dal commit al codice in produzione
- **Mean Time to Recovery (MTTR)**: Quanto velocemente ripristiniamo dopo un incidente
- **Change Failure Rate**: Percentuale di deploy che causano failure in produzione

### Security Metrics: Sicurezza Quantificabile

La sicurezza non può essere un'impressione, deve essere misurabile:

**Vulnerabilities Management**:

- Mean Time to Detection (MTTD): Tempo per scoprire una vulnerabilità
- Mean Time to Remediation (MTTR): Tempo per fixare dopo la scoperta

> - Vulnerability Density: Numero di vulnerabilità per 1000 linee di codice

Reviewer: Non è sempre su 1000 righe di codice, ma dipende dall'unità di misura che vogliamo assegnargli

- Critical vs High vs Medium: Distribuzione per severità

**Compliance Score**:

- Percentage of compliant deployments: Deploy che passano tutti i security gates
- Policy violations per sprint: Trending di violazioni nel tempo
- Security debt ratio: Vulnerabilità note non ancora risolte vs nuove feature

**Security Coverage**:

- Percentage di codice coperto da SAST/DAST
- Dipendenze scansionate vs totale
- Secret scanning hit rate e false positive rate

### DevEx Metrics: L'Esperienza Quantificata

> Developer Experience non è solo sentiment, ma può essere misurata oggettivamente:

Reviewer: Rivedere la frase, "sentiment" non ha senso

**Developer Satisfaction (Survey-based)**:

> - Net Promoter Score (NPS) del team

Reviewer: spiegare cos'è

- Tool satisfaction rating per ogni strumento nella toolchain
- Perceived productivity score (self-reported)
- Frequenza ideale: Quarterly pulse survey

**Onboarding Time**:

- Time to first commit: Dal day 1 al primo codice in repository
- Time to first production deploy: Quando un nuovo developer deploya autonomamente
- Time to independence: Quando non necessita più mentoring costante
- Benchmark: < 1 settimana per first commit, < 1 mese per independence

**Context Switching Metrics**:

> - Number of tools used daily: Meno è meglio (ideale < 5)

Reviewer: Non sarei d'accordo sul numero, nel senso, non si può dare un numero "assoluto". Già solo considerato un processo che include l'IDE di sviluppo, con integrazioni varie, l'uso di un database, una board Jira/Trello o quel che sia, le pipeline di GitHub/GitLab o i job su Jenkins, Docker per il deploy con annesso registry, strumenti per lo scan di sicurezza siamo fuori

- Average interruptions per day: Meeting, incident, support requests
- Focus time blocks: Ore consecutive senza interruzioni
- Work In Progress (WIP) items: Task parallele per developer

> ### Metriche Operative Aggiuntive

Reviewer: Aggiuntive nel senso di opzionali o rispetto a quanto detto sopra? Non è chiaro

Author: Tolte perchè forse sono anche troppe.

**Build & Pipeline Health**:

- Build success rate: Percentage di build che completano con successo
- Average build time: Durata media della pipeline end-to-end
- Flaky test rate: Test che falliscono intermittentemente
- Pipeline bottlenecks: Stage che rallentano il flusso

**Code Quality Trends**:

- Code coverage trajectory: Trending della copertura test
- Technical debt ratio: Tempo stimato per fix vs nuovo sviluppo
- Code review turnaround: Tempo medio per review approval
- Documentation freshness: Ultima modifica doc vs codice correlato

**Alert e Threshold**:

- MTTR > 4 ore: Alert immediato
- Build success rate < 80%: Review processo richiesto
- Vulnerability critical non fixata > 24h: Escalation automatica
- Developer NPS < 7: 1-on-1 meeting triggered

## Conclusione

> Il DevOps moderno non è più opzionale: è la baseline per esistere nel mercato software contemporaneo. Non è una destinazione ma un viaggio di miglioramento continuo.

Reviewer: Questo passaggio non mi torna. Forse sarebeb meglio dire che "la filosofia DevOps moderna è alla base di tutto il software con cui lavoriamo oggi, e pertanto adattarsi al cambiamento e seguire questi principi è fondamentale per crescere in questo ruolo"

## Bibliografia e Fonti

1. **Atlassian CALMS Framework**: https://www.atlassian.com/devops/frameworks/calms-framework - Framework di valutazione DevOps
2. **NIST DevSecOps Practices**: https://www.nccoe.nist.gov/projects/secure-software-development-security-and-operations-devsecops-practices - Standard di sicurezza
3. **Scaled Agile Framework DevOps**: https://framework.scaledagile.com/devops - Definizioni e pratiche DevOps
4. **Cprime CALMS Analysis**: https://www.cprime.com/resources/blog/defining-the-5-devops-principles-calms/ - Analisi dettagliata principi CALMS
5. **Microsoft Developer Experience**: https://developer.microsoft.com/en-us/developer-experience - Definizione e framework DevEx
6. **Splunk DevEx Guide**: https://www.splunk.com/en_us/blog/learn/developer-experience-devex.html - Guida completa DevEx
7. **DevOps.com Security Best Practices**: https://devops.com/securing-the-devops-pipeline-tools-and-best-practices-2/ - Best practices sicurezza 2024
8. **Cortex DevOps Security Guide**: https://www.cortex.io/post/devops-security-best-practices - Strategie security moderne

## Disclaimer

> Correzione ortografica e sintattica gentilmente offerta da una AI 🤖✍️: io ci ho messo le idee 💡, lei i congiuntivi 📚.

Reviewer 3: Grazie per la correttezza, aggiungeremo un apposito disclaimer in cima come fatto per altri contenuti!!
